### ISSUE TITLE
add settings to disable "Speculative pre-connections"

### ISSUE THREAD

**Issue by Radagast** [RequestPolicyContinued/requestpolicy#628]
Currently I'm using RPC (default deny) and NoScript but I still see requests to locations like:

> edge-star-shv-01-fra3.facebook.com	https 	Established	

Which appears to originate from the 'social' buttons on any given page. In RPC I have facebook (and other social networks) blocked but the connection mentioned above still occurs and doesn't show in the request log. 

I can block these connections using something like adblock or ublock but I'd rather not add anything else. Is there a way, using RPC,  to completely block all requests from buttons for all social networks?

**Comment by myrdd** [comment]
Please provide more information for being able to reproduce the problem. For example, on which site can you see those requests? Also, where (in what program) did you see that a request to `edge-star-shv-01-fra3.facebook.com` has been established?
Yes, normally you should be able to block requests to other domains, caused by social media buttons.

**Comment by Radagast** [comment]
Thanks for the reply. This might actually be something I'm doing, however...

For this example, the site in question is - poe.garena.ru. Once the site loads there are several 'Social' buttons at the top of the page. If I hover over the buttons a connection is established with the site associated with the button. This is something that can be seen in any network monitoring program such as netstat or a program such as [CurrPorts](http://www.nirsoft.net/utils/cports.html)

In RPC I have block rules for vk and facebook but I'm guessing the 'hovering' over the buttons is seen as a direct connection and not a cross site request?

**Comment by myrdd** [comment]
I could not reproduce this right now. I've set milestone to beta11 so that I don't forget it. Finishing beta10 is really urgent right now (should be released before Firefox 38), I hope you understand that.

Maybe someone else can check whether this issue can be reproduced?

**Comment by Radagast** [comment]
No problem. I've dropped back to 1.0.beta8.2 for now as I'm finding a few issues with beta 9+

**Comment by nodiscc** [comment]
Hi, I was not able to reproduce this using the latest beta, Debian Jessie, Firefox (Iceweasel) 37.0.1, and monitoring requests with Wireshark. I visited http://poe.garena.ru/, allowed requests to `cloudfront.net` for proper display, disabled my ad blocker, and hovered over the social buttons at the top. No requests except to `garena.ru` and `cloudfront.net` show up in Wireshark, over HTTP or TLS.

@Radagast is it possible another program/addon/setting is causing this? Can you do the following?

 * [Create a new, blank profile in Firefox](https://support.mozilla.org/en-US/kb/profile-manager-create-and-remove-firefox-profiles), with no addons installed.
 * Install the latest RequestPolicy Continued from https://github.com/RequestPolicyContinued/requestpolicy/releases
 * Navigate to your test page and try to reproduce the bug, and tell us if it happens again.

If nothing shows up in the Request log, it may be caused by another addon in the background. If you're using µBlock, you can also open it's request log and try to find the request to `edge-star-shv-01-fra3.facebook.com` when you load poe.garena.ru.

**Comment by Radagast** [comment]
After testing, it seems this 'problem' is related to firefox 38 (I haven't looked at Nightly yet) Steps taken:

First:
1. fx 37 with clean profile, RPC set to default deny with no subscriptions 
2. Load poe.garena.ru and allow cloudfront
3. Hover over the social buttons and observe network connections
4. No additional connections made from buttons

Second:
1. fx 38 with clean profile, RPC set to default deny and no subscriptions 
2. Load poe.garena.ru and allow cloudfront
3. Hover over the social buttons and observe network connections
4. Connections are made from buttons to facebook and others.

![Imgur](http://i.imgur.com/jhGbf3M.jpg)

**Comment by acrobat1** [comment]
Do you happen to have **network.predictor.enabled**  set on true in Firefox beta?

**Comment by Radagast** [comment]
In fx 37 the default for network.predictor.enabled is false. In fx 38 the default value is true. I was unaware Mozilla had renamed seer to predictor and hadn't updated my user.js with the new names. I've also just read [Bug 1005566](https://bugzilla.mozilla.org/show_bug.cgi?id=1005566) Changing the value to false does seem to fix the issue...

**Comment by acrobat1** [comment]
>In fx 37 the default for network.predictor.enabled is false. In fx 38 the default value is true.

That's right, maybe an option to toggle it off in RP, in line with what RP already has for link prefetching and DNS prefetching may be added in the future?

**Comment by nodiscc** [comment]
Well spotted! The documentation about Firefox's network prediction features is very sparse, I only found this:

 * https://bugzilla.mozilla.org/show_bug.cgi?id=1005566
 * http://www.ghacks.net/2014/05/11/seer-disable-firefox/

We need to find out whether it is relevant to RPC's goals to block such requests. Does it leak information that can be used to track users? Does it make users vulnerable to attacks? The list of related preferences can be found by following the bugzilla link above.

**Comment by acrobat1** [comment]
>We need to find out whether it is relevant to RPC's goals to block such requests. Does it > leak information that can be used to track users?

Well, that sounds like a no-brainer for me: if by just hovering on a "social" button or similar widget (which may well happen unintentionally, by simply moving the mouse around) Firefox contacts a webserver, that's in principle not so different from DNS prefetching and it can only help tracking ;)

On a side note, Mozilla has been adding so many "services" each with its own server (just have a quick look in about:config at things like *services.push.enabled* or *browser.selfsupport.url* ) that an extension may be needed just to disable all such services sending packets in the background...

**Comment by Radagast** [comment]
I agree with acrobat1. This particular pref really could be considered a candidate for tracking and, in my opinion, should be off by default.

**Comment by myrdd** [comment]
I also agree. Thank you @nodiscc for changing the title of this issue.

@acrobat1 about such services: I think as long as they go through the `ContentPolicy` (see [`nsIContentPolicy`](https://developer.mozilla.org/en-US/docs/Mozilla/Tech/XPCOM/Reference/Interface/nsIContentPolicy)) and get therefore checked by RPC I see no need to disable those services. But if you know a service that RPC doesn't catch, I'm happy if you tell me! Btw one such issue is #586.

**Comment by acrobat1** [comment]
I didn't know that RP could  also block such services, I can remember that not so long ago I've spotted packets from *push.mozilla.com* and *tracking.services.mozilla.com* , I don't remember at which version RP was back then (let's say about 4-5 months ago) - but I had to disable those services in about:config to get rid of that traffic.

**Comment by nodiscc** [comment]
Thisis a very cool project https://github.com/pyllyukko/user.js/. It provides a hardened [user.js] Firefox configuration file, with each setting clearly documented. Maybe RPC's advanced prefs `about:requestpolicy?advancedprefs` dialog could be extended to show switches for some of these prefs (enable/disable WebRTC https://github.com/RequestPolicyContinued/requestpolicy/issues/586, Prefetching, OCSP, prediction https://github.com/RequestPolicyContinued/requestpolicy/issues/628...)

@acrobat1 do you have more info about requests to _push.mozilla.com_ and _tracking.services.mozilla.com_? You can open another issue

@pyllyukko ping and nice work! What settings do you think should be added for easy access in the addon preferences?

**Comment by myrdd** [comment]
> I didn't know that RP could also block such services

I think there are several requests in the background that are not displayed in the RP menu. But yes, RP is able to block them, as long as they go through the Content Policy, which is the normal case.

@nodiscc looks good! thanks for the link!

> Maybe RPC's advanced prefs dialog could be extended to show switches for some of these prefs

I agree. Collecting them could be another task / issue.

**Comment by NinjaSentry** [comment]
Ideally Request Policy should be like a firewall, with the ability able to detect and provide the option to allow/block all background requests, from any website and from within the browser itself.

**Comment by myrdd** [comment]
> provide the option to allow/block all background requests, from any website and from within the browser itself

@NinjaSentry see #499 for that.

**Comment by myrdd** [comment]
@Radagast could you please check if setting `network.http.speculative-parallel-limit` to `0` solves this issue as well, even if `network.predictor.enabled` is `true`? Thats's at least suggested on [this Mozilla support page](https://support.mozilla.org/en-US/kb/how-stop-firefox-making-automatic-connections#w_speculative-pre-connections).

**Comment by myrdd** [comment]
If I understand / have tested correctly, then:

* `network.http.speculative-parallel-limit=0` implies `network.predictor.enabled=false`
* The "predictor", disabled by `network.predictor.enabled=false`, is just _one_ speculative mechanism. It makes requests when hovering links, for example.
* Setting `network.http.speculative-parallel-limit` to `0` will disallow _all_ speculative mechanisms.

Looks like `network.predictor.enabled=false` does _not_ disable `<link rel="preconnect">` [1] connections. I used `tcpdump` on debian/linux.
[1] [Source code commit](https://hg.mozilla.org/mozilla-central/rev/020545d52adf), [doc](https://developer.mozilla.org/en-US/docs/Web/HTML/Link_types), [release note](https://developer.mozilla.org/en-US/Firefox/Releases/39#HTML)

So I'll make a setting for "Speculative pre-connections", just as I described [on the wiki](https://github.com/RequestPolicyContinued/requestpolicy/wiki/What-is-prefetching).

**Comment by Radagast** [comment]
@myrdd.
From what I've seen over the last few months, setting network.http.speculative-parallel-limit to 0 alone is not enough, it's also important to network.predictor.enabled=false. Personally, I also have settings in about:config to block all prefetch requests as well.

**Comment by myrdd** [comment]
Not sure what that means for RequestPolicy. I'd like to keep the number of "main" prefereces (such as prefetch prefs) handled by RP to a minimum – at least as long as we've got the current system "how they are handled" (*). I guess for now I'll only add the `network.http.speculative-parallel-limit` setting to the recommended prefs, and will consider changing the system itself at a later time.

(*) I've created issue https://github.com/RequestPolicyContinued/requestpolicy/issues/749. I'd be happy to hear your opinion about it.
